### PR TITLE
S3 support

### PR DESCRIPTION
**Pull request by theus77**

### CODE REVIEW

File: Controller/FileController.php [hunk 1]
```diff
@@ -91,17 +92,33 @@ private function getFile(Request $request, string $sha1, string $disposition): R
...
+                throw new NotFoundHttpException('Impossible to find the item corresponding to this id: '.$sha1);
+            }
+
+            $response = new StreamedResponse(
+                function () use ($handler) {
+                    while (!feof($handler)) {
+                        print fread($handler, 8192);
+                    }
+                }, 200, [
+//                'Content-Disposition' => $disposition.'; '.HeaderUtils::toString(array('filename' => $name), ';'),
```
Comment (theus77) [inline Controller/FileController.php]: I intentionally leave those comment as I think this part can be improved

File: Storage/StorageManager.php [hunk 2]
```diff
@@ -26,58 +25,91 @@ class StorageManager
...
      * @return string
      */
-    public function getCacheFile(string $hash, ?string $context = null): string
+    public function getFile(string $hash, ?string $context = null): string
     {
-        return $this->read($this->cacheAdapters, $hash, $context);
+        $resource = $this->read($this->adapters, $hash, $context);
+        $filename = tempnam(sys_get_temp_dir(), 'EMS');
+        file_put_contents($filename, $resource);
+        return $filename;
```
Comment (dameert) [inline Storage/StorageManager.php]: How temporary is your file ?

Consider http://php.net/manual/en/class.spltempfileobject.php

File: Storage/StorageManager.php [hunk 3]
```diff
@@ -117,17 +149,16 @@ public function getPublicImage(string $name): string
     }
 
     /**
-     * @param AdapterInterface[]|iterable $adapters
+     * @param StorageInterface[]|iterable $adapters
      * @param string                      $hash
      * @param string|null                 $context
      *
-     * @return string
+     * @return resource
```
Comment (dameert) [inline Storage/StorageManager.php]: Resource, but prefere to use : Resource

File: Controller/FileController.php [hunk 4]
```diff
@@ -46,84 +63,51 @@ public function view(Request $request, string $sha1)
      *
```
Comment (dameert) [inline Controller/FileController.php]: @deprecated for IDE in the docblock

File: Controller/FileController.php [hunk 5]
```diff
@@ -46,84 +63,51 @@ public function view(Request $request, string $sha1)
...
-     */
-    private function createResponse(string $sha1)
+    public function download(Request $request, string $sha1)
     {
-        try {
-            $file = $this->storageManager->getFile($sha1);
-        } catch (NotFoundException $ex) {
-            throw new NotFoundHttpException('file not found');
-        }
+        @trigger_error("FileController::download is deprecated use the ems_asset twig filter to generate the route", E_USER_DEPRECATED);
```
Comment (dameert) [inline Controller/FileController.php]: add @deprecated in docblock

File: Controller/FileController.php [hunk 6]
```diff
@@ -46,84 +63,51 @@ public function view(Request $request, string $sha1)
...
     }
 
     /**
      * @param Request $request
      * @param string  $sha1
+     * @param string  $disposition
      *
-     * @return bool|Response
+     * @return Response
      */
```
Comment (dameert) [inline Controller/FileController.php]: add @deprecated in docblock

File: Entity/AssetStorage.php [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    public function setContext($context): AssetStorage
+    {
+        $this->context = $context;
+        return $this;
+    }
+
+    /**
+     * @return bool
+     */
+    public function isConfirmed(): bool
```
[the remark this answers is not in the document]
Comment (dameert) [reply Entity/AssetStorage.php]: Please clarify

File: Storage/Service/AbstractUrlStorage.php [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+    /**
+     * @param string $hash
+     * @param null|string $context
+     * @return \DateTime|null
+     */
+    public function getLastUpdateDate(string $hash, ?string $context = null): ?\DateTime
+    {
+        $path = $this->getPath($hash, $context);
+        if (file_exists($path)) {
+            $time = @filemtime($path);
```
[the remark this answers is not in the document]
Comment (dameert) [reply Storage/Service/AbstractUrlStorage.php]: Remove the @

File: Storage/Service/AbstractUrlStorage.php [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+    /**
+     * @param string $hash
+     * @param null|string $cacheContext
+     * @return int
+     */
+    public function getSize(string $hash, ?string $cacheContext = null): ?int
+    {
+        $path = $this->getPath($hash, $cacheContext);
+        if (file_exists($path)) {
+            return @filesize($path);
```
[the remark this answers is not in the document]
Comment (dameert) [reply Storage/Service/AbstractUrlStorage.php]: remove the @

File: Storage/Service/SftpStorage.php [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    protected function getBaseUrl(): string
+    {
+
+        if (!function_exists('ssh2_connect')) {
+            throw new Exception("PHP functions Secure Shell are required by $this. (ssh2)");
+        }
+
+        if (!$this->connection) {
+            $this->connection = @ssh2_connect($this->host, $this->port);
```
[the remark this answers is not in the document]
Comment (dameert) [reply Storage/Service/SftpStorage.php]: remove  @

File: Storage/Service/SftpStorage.php [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+            if (!$this->connection) {
+                throw new Exception("Could not connect to $this->host on port $this->port.");
+            }
+            ssh2_auth_pubkey_file($this->connection, $this->username, $this->publicKeyFile, $this->privateKeyFile, $this->passwordPhrase);
+        }
+
+        if (!$this->sftp) {
+            $this->sftp = @ssh2_sftp($this->connection);
```
[the remark this answers is not in the document]
Comment (dameert) [reply Storage/Service/SftpStorage.php]: remove @

**Pull request merged**